src/Agent.Worker/Build/BuildDirectoryManager.cs

```diff
@@ -144,6 +145,78 @@ public sealed class BuildDirectoryManager : AgentService, IBuildDirectoryManager
                     newConfig.RepositoryType = repository.Type;
                 }
 
+                // Make sure repository in the right location.
+                if (string.IsNullOrEmpty(newConfig.RepositoryDirectory))
+                {
+                    newConfig.RepositoryDirectory = newConfig.SourcesDirectory;
+                }
+
+                string requiredRepositoryPath;
+                var path = repository.Properties.Get<string>(RepositoryPropertyNames.Path);
+                if (!string.IsNullOrEmpty(path))
+                {
+                    if (path.IndexOfAny(Path.GetInvalidPathChars()) > -1)
+                    {
+                        throw new InvalidOperationException($"{path} contains invalid path characters.");
+                    }
+                    else if (Path.GetFileName(path).IndexOfAny(Path.GetInvalidFileNameChars()) > -1)
+                    {
+                        throw new InvalidOperationException($"{path} contains invalid folder name characters.");
+                    }
+                    else if (path.Contains($"..{Path.DirectorySeparatorChar}") || path.Contains($"..{Path.AltDirectorySeparatorChar}"))
+                    {
+                        throw new InvalidOperationException($"{path} can not contains relative path.");
+                    }
+                    else if (Path.IsPathRooted(path))
+                    {
+                        throw new InvalidOperationException($"{path} can not be a rooted path.");
+                    }
+                    else
+                    {
+                        requiredRepositoryPath = Path.Combine(newConfig.SourcesDirectory, path);
```

Comment:
What about removing segments like `.`? Note, sourceproviderbase.cs has similar logic (and unit tests iirc)

Someone will set a path like `./`, which will cause a problem somewhere